src/main/java/frc/robot/subsystems/SwerveSubsystem.java

```diff
@@ -0,0 +1,190 @@
+        MotorConfig.BackRightModuleTurn,
+        AbsoluteEncoders.BackRightModule,
+        "BR");
+
+    private AHRS gyro = new AHRS(SPI.Port.kMXP);
+    private SwerveDriveOdometry odometer = new SwerveDriveOdometry(DriveConstants.kDriveKinematics, new Rotation2d(0));
+
+    private DataLog datalog = DataLogManager.getLog();
+    private DoubleLogEntry translationXOutputLog = new DoubleLogEntry(datalog, "/swerve/txout"); //Logs x translation state output
+    private DoubleLogEntry translationYOutputLog = new DoubleLogEntry(datalog, "/swerve/tyout"); //Logs y translation state output
+    private DoubleLogEntry rotationOutputLog = new DoubleLogEntry(datalog, "/swerve/rotout"); //Logs rotation state output
+    private BooleanLogEntry controlOrientationLog = new BooleanLogEntry(datalog, "/swerve/orientation"); //Logs if robot is in FOD/ROD
+
+    public boolean controlOrientationIsFOD;
+
+    public SwerveSubsystem() {
+        new WaitUntilCommand(this::gyroReady)
+        .andThen(new InstantCommand(this::zeroHeading,this))
+        .schedule();
+        controlOrientationIsFOD = true;
+    }
+
+    private boolean gyroReady() {
+        return !gyro.isCalibrating();
+    }
+
+    public void zeroHeading() {
+        gyro.reset();
+    }
+
+    public double getHeading() {
+        return Math.IEEEremainder(gyro.getAngle(), 360); //Clamps angle output between -180 and 180 degrees
+    }
+
+    public Rotation2d getRotation2d() {
+        return Rotation2d.fromDegrees(getHeading());
+    }
+
+    public Pose2d getPose() {
+        return odometer.getPoseMeters();
+    }
+
+    public void resetOdometry(Pose2d pose) {
+        odometer.resetPosition(pose, getRotation2d());
+    }
+
+    public void toggleOrientation(){
+        //Toggle control orientation from FOD/ROD
+        controlOrientationIsFOD = !controlOrientationIsFOD;
+        controlOrientationLog.append(controlOrientationIsFOD);
+    }
+
+    public SwerveModuleState[] convertToModuleStates(double xTranslation, double yTranslation, double rotation) {
+        //Takes axis input from joysticks and returns an array of swerve module states
+
+        double x = yTranslation; //Intentional, x in swerve kinematics is y on the joystick
+        double y = xTranslation;
+        double r = rotation;
+
+        //Map to speeds in meters/radians per second
+        x *= DriveConstants.kPhysicalMaxSpeed;
+        y *= DriveConstants.kPhysicalMaxSpeed;
+        r *= DriveConstants.kPhysicalMaxAngularSpeed;
+
+        //Log speeds used to construct chassis speeds
+        translationXOutputLog.append(x);
+        translationYOutputLog.append(y);
+        rotationOutputLog.append(r);
+
+        //Construct Chassis Speeds
+        ChassisSpeeds chassisSpeeds;
+        if(controlOrientationIsFOD){
+            //Field Oriented Drive
+            chassisSpeeds = ChassisSpeeds.fromFieldRelativeSpeeds(x, y, r, this.getRotation2d().plus(new Rotation2d((r * Robot.kDefaultPeriod)/2)));
```

Comment:
We should comment why this is here.  I won't remember.  
And add a TODO to check and see if it works on the physical hardware